### ISSUE TITLE
Wiki Documents

### ISSUE THREAD

**Issue by anastassia-b** [youknowhu/Hippocampus#1]
Hey Kimmy,

Awesome work on the wiki for your full stack proposal! I've had the chance to review it and here is my feedback:

# Wiki Page Home

- [x] Is the first page you see upon entering the wiki
- [x] Contains a welcome message
- [x] Contains a link/placeholder for a link to the live page
- [x] All links in the right sidebar should contain each wiki page and link to the correct page
- [x] Correctly formatted
    - [x] each wiki page is listed in bullet points
    - [x] all links route the correct page

### Comments:

Looks great to me!

# MVP List

- [x] Correctly formatted
- [x] Should have 7 MVPs
    - [x] 3 of those are User Auth, Heroku, and Production README.
    - [x] The other 4 are from the [MVP List](https://github.com/appacademy/curriculum/blob/master/full-stack-project/proposal/mvp-list.md) or otherwise clarified
- [x] At least one or two CRUD features:
  - [x] State what CRUD operations are planned (creation, reading, updating, deletion)
- [x] Estimates how long it will take to code each MVP

### Comments:

- [x] Will users be able to delete Reviews or Bookings? Please make the CRUD operations explicit.
- [x] Add the bullet: "Users can see a homepage of listings" to the Listing MVP.
- When you later open each MVP as an issue, in the issue description please have the following bullets for each app-specific MVP:
  + Adequate styling
  + Smooth, bug-free navigation
  + Seeded with info to demonstrate this feature

# Database Schema

- [x] Contains correct datatypes
- [x] Contains appropriate constraints/details
    - [x] primary key
    - [x] not null
    - [x] unique
    - [x] indexed
    - [x] foreign key
- [x] Contains bullet points after the table that state which foreign keys will reference to which table, or references to the associations which will be made
- [x] Correctly formatted

### Comments:

The user's image_url is not null. In order to make the account sign up process easier, you can have a default image url that gets assigned when you try to make a new user (similar to how we ensure_session_token). That way someone doesn't have to upload a photo when making a new account, but they will get a default photo they can later change (as bonus).

Well done!

# Routes

- [x] Routes accurately reflect MVPs and Component Hierarchy
- [x] Contains the following sections: HTML, API Endpoints(Backend), and Frontend Routes
- [x] Each route has a description
- [x] Covers the frontend functionality of the site through routes that match to the wireframes
- [x] API Endpoint routes contains wildcard variables written in `snake_case`
- [x] Frontend routes contains wildcard variables written in `camelCase`
- [x] Routes does not contain [superfluous routes](https://github.com/appacademy/curriculum/blob/master/rails/readings/routing-part-ii.md)
- [x] Have API routes that will allow the front end to get all info it needs and does not have unneeded routes
- [x] Correctly formatted

### Comments:

- [x] Probably don't need: `GET /api/listings/:listing_id/reviews` - the reviews will likely come along when you GET a specific listing (just like the listing photos).

# Sample State

- [x] State shape is flat!
- [x] State's keys are camelCased
- [x] All keys within the values in the state are accessible in the schema (Ex: if you create a `img_url` key in the state, make sure your database also has a column that contains an `img_url`)
- [x] Correctly formatted
    - [x] Sample state is rendered with 3 back ticks before and after. This will display the state as a code block instead of a giant line of text
    - [x] Top level slices
        - [x] `entities`
        - [x] `session`
        - [x] `errors` (here or in `ui`)
        - [x] `ui` (if needed)

### Comments:

- [x] In addition to `listing_photos_ids: [1, 3, 5, 10, 15]` , a listing will need `listing_review_ids` to key into the reviews slice of state.
- [x] The sample state does not have a `listing_photos` slice of state.
- [x] You may want to store a listing `icon_url` with each specific listing on the frontend. That way when displaying the listings on the homepage or search, a listing can have an icon without having to grab all of the listing_photos from the backend. This could be added in json to the `GET/listings`, pulling the first image via the listing_photos association.

Note: A user profile page is not part of the MVPs. If it was, each user would have a `booking_ids` array so past and upcoming bookings could be displayed on their page. I think the bookings slice of state how it is now is good-- it could allow you to show on a specific listing: `already booked for X date!` for the current user.

# Component Hierarchy with Wireframes

- [x] All routes listed in the routes wiki page is also listed in the component hierarchy
- [x] Each component, if necessary, has route, state, and other components it renders [example](https://github.com/appacademy/bluebird/wiki/component-hierarchy)
- [x] Necessary components are wrapped in a container
- [x] Correctly formatted
- [x] Wireframes:
  - [x] Every component listed in the component hierarchy wiki page has at least one wireframe
  - [x] Every MVPs listed on the MVP wiki page has at least one wireframe
  - [x] Every route listed on the route wiki page has at least one wireframe
  - [x] Contains Containers
  - [x] Contains Lists
  - [x] Contains ListItems

### Comments:

* I don't think you need `:dateParams` in the frontend route for bookings?
* There is a search component on the homepage and a search in the navbar of the search page. If you make the search bar a very lightweight component, you can reuse it on the homepage, search page, and potentially in every navbar. Then you can wrap it in a Higher-Order-Component on the homepage to give it extra functionality/design.

---

Overall excellent proposal! Please fix these small comments by tomorrow prior to starting coding. Thank you!

**Comment by youknowhu** [comment]
Hi @anastassia-b, 

Thanks for your feedback -- it was extremely helpful. I have completed all of the outstanding items you identified. 

A few follow up notes and questions:

**1/ Routes** -- I added few additional routes related to reviews to ensure I had both the front-end and back-end routes to support the full CRUD cycle for reviews. 

- Updated back-end routes: 
   - `GET /api/reviews/:id` - return specific review
   - `PATCH /api/reviews/:id` - update review
   - `DELETE /api/reviews/:id` - delete review

- Updated front-end route: 
  - `/listings/:listingId/:reviewId/edit `- edit review

**2/ Sample State** 
- Similar to the `listingReviewIds`, I decided to add a `listingBookingIds` array so that when we load a listing we have everything we need (instead of going through all of bookings. 


**3/ React Component Hierarchy & Wireframes** 
- This is the section I have the most questions on, so I would love to dig into both of your comments a little further. I also have a few additional questions of my own, that I would love to discuss further live!


Thanks so much!

Kimmy
